pandas/core/ops/mask_ops.py

```diff
@@ -176,3 +177,55 @@ def kleene_and(
 def raise_for_nan(value, method):
     if lib.is_float(value) and np.isnan(value):
         raise ValueError(f"Cannot perform logical '{method}' with floating NaN")
+
+
+def sum(
+    values: np.ndarray, mask: np.ndarray, skipna: bool, min_count: int = 0,
+):
+    """
+    Sum for 1D masked array.
+
+    Parameters
+    ----------
+    values : np.ndarray
+        Numpy array with the values (can be of any dtype that support the
+        operation).
+    mask : np.ndarray
+        Boolean numpy array (True values indicate missing values).
+    skipna : bool, default True
+        Whether to skip NA.
+    min_count : int, default 0
+        The required number of valid values to perform the operation. If fewer than
+        ``min_count`` non-NA values are present the result will be NA.
+    """
+    if not skipna:
+        if mask.any():
+            return libmissing.NA
+        else:
+            if _below_min_count(values, None, min_count):
+                return libmissing.NA
+            return np.sum(values)
+    else:
+        if _below_min_count(values, mask, min_count):
+            return libmissing.NA
+
+        if _np_version_under1p17:
+            return np.sum(values[~mask])
+        else:
+            return np.sum(values, where=~mask)
```

Comment:
out of curiosity, is this more efficient than the version used in older numpy?

Reply:
Good question. With, with the array in my top post it gives a slight boost (a 5-8% speed-up), but it varies quite a bit (I could also have it slower with more np.nan values). 
But when going to larger arrays, there seems to be a clearer difference:

```
In [40]: a = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, np.nan]*1_000_000)   

In [41]: mask = np.isnan(a)  

In [42]: %timeit np.sum(a, where=~mask)    
17.7 ms ± 861 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [43]: %timeit np.sum(a[~mask])  
43.3 ms ± 4.13 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Although the difference gets smaller when there are more missing values (eg with 50% missing values instead of 10% in the above, it becomes 25 vs 32ms). 

But I assume there is also a memory benefit (avoiding the temporary array), although I am not fully familiar with the inner workings of this in numpy.